create new crate for core-ops

jandremarais opened the pull request:
Separate the core tensor operations from the core crate to a core-ops crate.

jandremarais commented:
@edgarriba do you want the operations currently defined in `kornia-core/src/tensor.rs` like `add`/`mul`/`powi`/etc to live in `kornia-core-ops/src/ops.rs`? 
If so, how should we keep the ability to call for example `tensor.powi(2)` since we can't define the impl for a foreign type? Do we need a wrapper struct?

jandremarais commented:
Ok. Should I start with something simple that is not yet implemented or something that is already implement for Tensor?

edgarriba commented:
maybe the norm operator needed by cosine distance

jandremarais commented:
I started with the norm operator but then realised it needed the `sum` reduce operator. So I did its implementation first.

edgarriba commented on crates/kornia-core-ops/src/ops.rs:
```diff
@@ -0,0 +1,111 @@
+
+/// Compute the sum of the elements in the tensor along dimension `dim`
+///
+pub fn sum<T, const N: usize, A>(
```
Maybe call it `sum_elements` to be more specific

jandremarais commented:
Marked this as ready for review so that the new core-ops crate can be merged without being blocked by all the ops that need to be implemented. I you would rather want this to be a bigger PR with more ops added first, I can change the status to draft and work on it further.

edgarriba commented on crates/kornia-core-ops/src/ops.rs:
```diff
@@ -0,0 +1,111 @@
+    }
+
+    #[test]
+    fn test_sum_2d() -> Result<(), TensorError> {
```
for sanity check can we test with a different types e.g. here `f32`?

edgarriba commented on crates/kornia-core/src/tensor.rs:
```diff
@@ -381,6 +381,26 @@ where
+    /// # Returns
+    ///
+    /// The list of indices to get the element from.
+    pub fn get_index(&self, offset: usize) -> [usize; N] {
```
if guess this should return an error if `offset` is larger than the number of elements

jandremarais replied:
Yes, I was wondering if this should follow the same pattern as the `get_iter_offset` method with a checked and unchecked version where the checked version returns an `Option`. What do you think?

edgarriba replied:
good idea, add test to check forth and back

edgarriba commented on crates/kornia-core-ops/src/error.rs:
```diff
@@ -0,0 +1,9 @@
+pub enum TensorOpsError {
+    /// Dimension out of bounds.
+    #[error("Dimension out of bounds. The dimension {0} is out of bounds ({1}).")]
+    DimOutOfBounds(usize, usize),
```
can we reuse or unify somehow with https://github.com/kornia/kornia-rs/blob/main/crates/kornia-core/src/tensor.rs#L23 ?

jandremarais replied:
I can add `DimOutOfBounds` to `TensorError`. Feels like it means something different to `IndexOutOfBounds`. The latter is when the index is greater than the size of the array and the former when the queried dimension/axis does not exist.

jandremarais replied:
I couldn't think of a scenario where this error is applicable outside of tensor operations (tensor indexing is already validated by the const generics) and therefore I created it in the core-ops crate initially.

edgarriba replied:
i see, yes makes sense. Make sure the description and error message is clear for both then

edgarriba commented on crates/kornia-core-ops/src/ops.rs:
```diff
@@ -0,0 +1,111 @@
+use crate::error::TensorOpsError;
+
+/// Compute the sum of the elements in the tensor along dimension `dim`
+///
```
add args in the docs and a small snippet example

edgarriba commented on crates/kornia-core-ops/src/ops.rs:
```diff
@@ -0,0 +1,111 @@
+
+/// Compute the sum of the elements in the tensor along dimension `dim`
+///
+pub fn sum_elements<T, const N: usize, A>(
```
for future PR -- worth finding an abstraction like here https://github.com/kornia/kornia-rs/blob/main/crates/kornia-core/src/tensor.rs#L710

one more thought is how to support with the same interface tensors with contiguous memory that do not require computing the offset for the given index.

edgarriba replied:
and something I coudn't figure out yet is how to add a `keepdim` e.g to reduce and return the tensor rank to N -1. @emilmgeorge maybe you have some thoughts here

emilmgeorge replied:
@edgarriba 
I think we can simply remove the `dim` dimension from `shape` and `strides` when `keepdim == false` since the num elements stay the same. I hope I'm not misunderstanding what you meant.

edgarriba replied:
this is actually the only limitation we have with generic tensor ranks. Hopefully we con figure out a workaround as this comes useful to convert tensor to onnxruntime or any other dl engine :)

see: https://github.com/kornia/kornia-rs/blob/8cd3394ab7aad40f0d1ef503366eca7a98756b94/crates/kornia-dnn/src/rtdetr.rs#L158

edgarriba replied:
where BTW, we should be able to bindo directly to cuda at some point

emilmgeorge replied:
Would it be feasible to make dimensions dynamic? Like `Tensor<T, A>` instead of `Tensor<T, N, A>` and then manage shape/stride at runtime using Vec. I don't know the whole code base well enough yet to know the possibility and consequences of this. Just sharing a thought. If feasible, it would solve this problem and give more flexibility. But I guess this would also cause some performance hit.

jandremarais replied:
I tried to experiment with this idea of keeping const generic ranks and using it to inform the sum operation on what rank tensor to return. I've added these traits heavily borrowed (simplified and probably worsen) from [dfdx](https://github.com/coreylowman/dfdx/blob/main/dfdx-core/src/shapes/shape.rs) with this [commit](https://github.com/kornia/kornia-rs/pull/142/commits/4b36377d7e122d3a6a9beb79bb7431ef6e1ce79b). Example usage can be seen in the test from this [commit](https://github.com/kornia/kornia-rs/pull/142/commits/d4fa9512851aa36bb230e47d619e7c60faadeb3d).

I might have messed up this PR now. Let me know what's best to do next.

edgarriba replied:
@jandremarais that's a nice feature ! Maybe as a separated PR we can introduce the reshape stuff, and later integrate into the ops?

edgarriba commented on crates/kornia-core-ops/src/ops.rs:
```diff
@@ -1,7 +1,76 @@
+{
+    fn reduce(self, dim: usize) -> Tensor<T, 2, A> {
+        let shape = match dim {
+            0 => [self.shape[1], self.shape[2]],
```
what if dim 0 is greater than 1 ? return error as it's not possible to perform the reduction ?

edgarriba commented:
@jandremarais any progress here ? we can land step by step and keep improving

edgarriba commented:
@jandremarais sound good -- consider this too https://github.com/apache/arrow-rs/issues/6492#issuecomment-2387214140. I might be working on a PR for it during the week. Arrow seems pretty unflexible, so maybe we should go for our full implementation

jandremarais commented:
@edgarriba I addressed all of your comments in the latest commits. Open to further feedback if I interpreted anything incorrectly or if there are any new suggestions.

edgarriba commented on crates/kornia-core-ops/src/ops.rs:
```diff
@@ -0,0 +1,152 @@
+
+    #[test]
+    fn test_sum_2d() -> Result<(), TensorOpsError> {
+        let data: [u8; 6] = [1, 1, 1, 1, 1, 1];
```
if one of this test can with non-equals numbers for sanity check, like
```suggestion
        let data: [u8; 6] = [1, 2, 3, 4, 5, 6];
```

edgarriba reviewed (approved):
LGTM. Make a rebase and all good !
